services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+import Control.Error (assertMay, failWith, failWithM)
+import Control.Lens (view, (?~), (^?))
+import Control.Monad.Except
+import Crypto.JWT hiding (params, uri)
+import Data.ByteString.Conversion
+import Data.Domain
+import Data.Id
+import Data.Misc (PlainTextPassword (PlainTextPassword))
+import qualified Data.Set as Set
+import Data.String.Conversions (cs)
+import Data.Text.Ascii
+import Data.Time
+import Imports hiding (exp)
+import OpenSSL.Random (randBytes)
+import Polysemy (Member)
+import Servant hiding (Handler, Tagged)
+import Wire.API.Error
+import Wire.API.OAuth as OAuth
+import qualified Wire.API.Routes.Internal.Brig.OAuth as I
+import Wire.API.Routes.Named (Named (..))
+import Wire.API.Routes.Public.Brig.OAuth
+import Wire.Sem.Jwk
+import qualified Wire.Sem.Jwk as Jwk
+import Wire.Sem.Now (Now)
+import qualified Wire.Sem.Now as Now
+
+--------------------------------------------------------------------------------
+-- API Internal
+
+internalOauthAPI :: ServerT I.OAuthAPI (Handler r)
+internalOauthAPI =
+  Named @"create-oauth-client" createNewOAuthClient
+
+--------------------------------------------------------------------------------
+-- API Public
+
+oauthAPI :: (Member Now r, Member Jwk r) => ServerT OAuthAPI (Handler r)
+oauthAPI =
+  Named @"get-oauth-client" getOAuthClient
+    :<|> Named @"create-oauth-auth-code" createNewOAuthAuthorizationCode
+    :<|> Named @"create-oauth-access-token" createAccessTokenWith
+    :<|> Named @"revoke-oauth-refresh-token" revokeRefreshToken
+    :<|> Named @"get-oauth-applications" getOAuthApplications
+    :<|> Named @"revoke-oauth-account-access" revokeOAuthAccountAccess
+
+--------------------------------------------------------------------------------
+-- Handlers
+
+createNewOAuthClient :: NewOAuthClient -> (Handler r) OAuthClientCredentials
+createNewOAuthClient (NewOAuthClient name uri) = do
+  unlessM (Opt.setOAuthEnabled <$> view settings) $ throwStd $ errorToWai @'OAuthFeatureDisabled
+  credentials@(OAuthClientCredentials cid secret) <- OAuthClientCredentials <$> randomId <*> createSecret
+  safeSecret <- liftIO $ hashClientSecret secret
+  lift $ wrapClient $ insertOAuthClient cid name uri safeSecret
+  pure credentials
+  where
+    createSecret :: MonadIO m => m OAuthClientPlainTextSecret
+    createSecret = OAuthClientPlainTextSecret <$> rand32Bytes
+
+    hashClientSecret :: MonadIO m => OAuthClientPlainTextSecret -> m Password
+    hashClientSecret = mkSafePassword . PlainTextPassword . toText . unOAuthClientPlainTextSecret
+
+--------------------------------------------------------------------------------
+
+getOAuthClient :: UserId -> OAuthClientId -> (Handler r) (Maybe OAuthClient)
+getOAuthClient _ cid = do
+  unlessM (Opt.setOAuthEnabled <$> view settings) $ throwStd $ errorToWai @'OAuthFeatureDisabled
+  lift $ wrapClient $ lookupOauthClient cid
+
+createNewOAuthAuthorizationCode :: UserId -> NewOAuthAuthorizationCode -> (Handler r) CreateOAuthCodeResponse
```

Comment:
I'm still a bit unhappy with the naming.  `NewOAuthAuthorizationCode` is a thing you need to create an `AuthorizationCode`?  This is [step 2-5](http://localhost:3000/developer/reference/oauth.html#supported-oauth-flow), correct?  And arrow 5 is not really going from wire-server to oauth client, but from wire-server to browser and then to oauth app by redirect, no?  Maybe add those steps from the docs here to the haddocks as well?

Reply:
Followed an inconsistent convention here with the naming. Will change to `CrearteXRequest`.